### ISSUE TITLE
CA1508 False Positive On Switch Expression with when keyword

### ISSUE THREAD

**Issue by Uzivatel919** [dotnet/roslyn-analyzers#5142]
### Analyzer

**Diagnostic ID**: [CA1508: Avoid dead conditional code](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1508)

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis),

**Version**: [SDK 5.0.300](https://dotnet.microsoft.com/download/dotnet/5.0)

**AND**

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 5.0.3 (Latest)

### Bug sample

```
static class Ca1508FalsePositiveOnSwitchExpression
{
	static void Example ()
	{
		string str = new (new [] {'A', 'B'});

		_ = str switch
		{
			string s when s.Length == int.MaxValue => "🝰", // CA1508: False Positive
		};

		_ = str switch
		{
			string s when s.Length is int.MaxValue => "🥛",  // CA1508: False Positive
		};

		_ = str switch
		{
			string { Length: int.MaxValue } => "🦧", // Ok
		};

		_ = str switch
		{
			string ( 1, _ ) => "🍌", // Ok
		};
	}

	static void Deconstruct ( this string str, out int length, out char[] chars )
	{
		length = str.Length;
		chars = str.ToCharArray ();
	}
}
```

### Expected behavior
No false positive.

### Actual behavior
Some false positive.
![CA1508 ELW](https://user-images.githubusercontent.com/26486200/120931674-f2091980-c6f2-11eb-8daf-d75a9537f364.png)

## Additional context
• error can suddenly disappear (try to change project type) from _Error List_ windows
• everytime can be found at build output
```
1>C:\Users\JC\source\repos\AlwayTruePatternMatching\Ca1508FalsePositiveOnSwitchExpression.cs(11,9,11,55): error CA1508: 'string s when s.Length == int.MaxValue => "🝰"' is always 'true'. Remove or refactor the condition(s) to avoid dead code.
1>C:\Users\JC\source\repos\AlwayTruePatternMatching\Ca1508FalsePositiveOnSwitchExpression.cs(16,9,16,55): error CA1508: 'string s when s.Length is int.MaxValue => "🥛"' is always 'true'. Remove or refactor the condition(s) to avoid dead code.
1>C:\Users\JC\source\repos\AlwayTruePatternMatching\Ca1508FalsePositiveOnSwitchExpression.cs(16,23,16,47): error CA1508: 's.Length is int.MaxValue' is always 'true'. Remove or refactor the condition(s) to avoid dead code.
```

[AlwayTruePatternMatching.zip](https://github.com/dotnet/roslyn-analyzers/files/6604491/AlwayTruePatternMatching.zip)

**Comment by mavasani** [comment]
This is partially addressed with https://github.com/dotnet/roslyn-analyzers/pull/6809, but still repros for second case, still generates the CA1508 false positive.
